tests/unit/config/test_config.py

```diff
@@ -0,0 +1,110 @@
+"""Unit test of config package functionality."""
+import configparser
+import logging
+import pytest
+import os
+
+import config
+from os.path import isfile
+
+logger = logging.getLogger(__name__)
+
+
+@pytest.fixture
+def preserve_config(scope="module"):
+    """Preserves the existing config file before running a function."""
```

Comment:
So you are testing on the existing config which the solution is using?  Could maybe this be alleviated by mocking?

Reply:
Maybe look at working on a tempfile.

`db_fd, db_path = tempfile.mkstemp()`

Ref.: https://flask.palletsprojects.com/en/2.0.x/testing.html

Reply:
Yes this uses the existing config file for testing. But before running the tests it moves it to `config.bak` while the tests are writing bad data to `config.ini`. After the tests, the config file is restored as it was originally. Initially i wanted to use a temp file for this, but this was hard to do because of OS specific paths.

Reply:
How so?  What issues did you come across?

Can’t you access the file using `db_path`?  (In the example that is)

Reply:
The problem is the `load_config()` function since it checks OS specific places for the config file. The reason it was written that way is to ensure you always get a valid configparser object back when you call `load_config()`. Even without a config file present you will get the default one generated in memory. When testing i could not find a clean method to override the OS specific path to the config file such that it becomes a temp file. I have never used any form of mocking, if you think it's a much better way of testing it feel free to submit a commit :smile: 

One option is to add an optional parameter to `load_config()` which overrides the path to the config file. This makes some tests easier since you can just store different config files in the `test_data` folder. However this won't test the functions that gets OS specific path to the config file.

Reply:
you make a new `load_config()` function under some testing namespace, and import and use it instead of the real `load_config()` function.

Reply:
The only part not tested in `load_config()` is the OS specific functions, see the above conversation. Otherwise is `load_config()` fully tested by taking backup of the possibly existing config file using the `preserve_config()` decorator. What i understood from @MindTooth is that instead of moving the file, i mock the functions used by the `ConfigParser` object.

Reply:
I need to look more into mocking myself.  The only reason I commented on this is that a test should never modify real data.  In this case, the actual config file that it used for production.  Not sure that testing the actual real life folders are necassery, but instead handle the error that might occur; permission, missing parent folder, etc.  How one would this proper, I don’t know atm.

Reply:
Implemented mocking when testing reading and writing in this commit da607cda2c9b14d7d9f9a89d96050f1eff6ef75a. Looking to do something similar to `test_read_config_garbage_data(...)` but this seems more complex to implement. Somehow i need to redirect the path used when calling `load_config()`. Any suggestions?

Reply:
This fixture is now removed with the deffbc336b078e9b5b10a92d03663a9af94ae697 commit. Modified `load_config()` to take in an optional path override. Tests use this in order to check malformed configuration files. Since the other functions have been changed to mock, the problematic fixture could be removed.

Reply:
![image](https://user-images.githubusercontent.com/33870508/126905650-6bfaab94-80e8-48fd-ae81-ccd3213b8c38.gif)